File: Algorithms/HackerEarth/LeafAndLimelightAttack.java
package HackerEarth;

import java.util.Scanner;

/**
 * Leaf and Limelight Attack
 * Algorithms  Dynamic Programming  Introduction to Dynamic Programming 1
 * Easy
 * @author tmosest
 *
 */
public class LeafAndLimelightAttack {
	
	private static long mod = (long) Math.pow(10, 9) + 9;
	private static int limit = (int) Math.pow(10, 7);
	private static long[] sumOfDiagnols = new long[limit + 1];
	
	public static void main(String[] args)
	{
		fillDiagnol();
		Scanner in = new Scanner(System.in);
		
		int qs = in.nextInt();
		
		for(int q = 0; q < qs; q++) {
			int n = in.nextInt();
			System.out.println(sumOfDiagnols[n]);
		}
		
		in.close();
	}
	
	private static void fillDiagnol()
	{
		sumOfDiagnols[0] = 0;
		sumOfDiagnols[1] = 1;
		sumOfDiagnols[2] = 10;
		for(int i = 3; i <= limit; i++) {
			sumOfDiagnols[i] = ((sumOfOutsideLayer(i) % mod) + (sumOfDiagnols[i - 2] % mod)) % mod;
		}
	}
	
	private static long sumDiagnols(int n)
	{
		if(n == 0)
			return 0;
		else if(n == 1)
			return 1;
		else if(n == 2) {
			System.out.println(2);
			return 10;
		}
		else if(sumOfDiagnols[n] != 0)
			return sumOfDiagnols[n];
		else {
			sumOfDiagnols[n] = ((sumOfOutsideLayer(n) % mod) + (sumDiagnols(n - 2) % mod)) % mod;
			return sumOfDiagnols[n];
		}	
	}
	
	private static long sumOfOutsideLayer(int n)
	{
		long sum = 0;
		long n2 = ((n % mod) * (n % mod)) % mod;
		for(int i = 0; i < 4; i++) {
			sum += (n2 - (((n - 1) % mod) * i) % mod) % mod;
			//System.out.println(sum);
		}
		return sum;
	}
}

File: Algorithms/HackerEarth/LeafAndLimelightAttack.c
/**
 * Leaf and Limelight Attack
 * Algorithms  Dynamic Programming  Introduction to Dynamic Programming 1
 * Easy
 * @author tmosest
 *
 */

#include <stdio.h>

#define MOD 1000000009L // (long) 10e9 + 9
#define LIMIT 10000000L // (int) 10e7

long long int sumOfDiagnols[LIMIT + 1];

void fillSumOfDiagnols();
long long sumOfOutsideLayer(int n);


/**
 *Get the party started
 */
int main()
{
	fillSumOfDiagnols();

	int qs;
	long long n;

	scanf("%d", &qs);

	for(int q = 0; q < qs; q++) {
		scanf("%lld", &n);
		printf("%lld\n", sumOfDiagnols[n]);
	}

	return 0;
}

/**
 * Function to fill the sumOfDiagnols in a D.P. style
 */
void fillSumOfDiagnols()
{
	sumOfDiagnols[0] = 0;
	sumOfDiagnols[1] = 1;
	sumOfDiagnols[2] = 10;
	for(int i = 3; i <= LIMIT; i++) {
		sumOfDiagnols[i] = ((sumOfOutsideLayer(i) % MOD) + (sumOfDiagnols[i - 2] % MOD)) % MOD;
	}
}

/**
 * Sums the four corners of an n by n spiral matrix
 */
long long sumOfOutsideLayer(int n)
{
	long long sum = 0;
	long long n2 = ((n % MOD) * (n % MOD)) % MOD;
	for(int i = 0; i < 4; i++) {
		sum += ((n2 - (((n - 1) % MOD) * i) % MOD) % MOD;
	}
	return sum;
}
